File: 10-regular-expression-matching/10-regular-expression-matching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
  const cache = new Map()
  const sl = s.length
  const pl = p.length
  const dp = (si, pi) => {
    if (pi === pl) return si === sl;
    if (si === sl) {
      if ((pl - pi)%2 !== 0) return false
      for (let i = pi+1; i < pl; i += 2) {
        if (p[i] !== '*') return false
      }
      return true
    }
    
    const key = si * 1000 + pi
    if (cache.has(key)) return cache.get(key)
    
    let result = false
    if (s[si] === p[pi] || p[pi] === '.') {
      if (pi < pl - 1 && p[pi+1] === '*')
        result = dp(si, pi + 2) || dp(si + 1, pi) 
      else
        result = dp(si + 1, pi + 1)
    } else if (pi < pl - 1 && p[pi+1] === '*') {
      result = dp(si, pi + 2)
    }
    
    cache.set(key, result)
    return result
  }
  return dp(0, 0)
}

File: 10-regular-expression-matching/10-regular-expression-matching.scala
import scala.collection.mutable.Map
object Solution {
  private val cache = Map.empty[Tuple2[Int, Int], Boolean]
  private def dp(s: String, si: Int, p: String, pi: Int): Boolean = {
    val sl = s.length
    val pl = p.length
    if (pi >= pl && si >= sl) return true
    if (pi >= pl && si < sl) return false
    if (si >= sl) {
      if ((pl-pi) % 2 != 0) return false
      else {
        for (i <- pi until pl) {
          if ((i-pi)% 2 > 0 && p.charAt(i) != '*') {
            return false
          }
        }
        return true
      }
    }
    
    val pair = (si, pi)
    if (cache contains pair) cache(pair) 
    else {
      val result: Boolean = 
        if (p.charAt(pi) == '.' || p.charAt(pi) == s.charAt(si)) {
          if (pi < (pl-1) && p.charAt(pi+1) == '*')
            dp(s, si, p, pi+2) || dp(s, si+1, p, pi)
          else dp(s, si+1, p, pi+1)
        }
        else {
          if (pi < (pl-1) && p.charAt(pi+1) == '*')
            dp(s, si, p, pi+2)
          else false
        }
      cache.put(pair, result)
      result
    }
  }
  def isMatch(s: String, p: String): Boolean = {
    cache.clear()
    dp(s, 0, p, 0)
  }
}